Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1366 @@
+      try self.invalidState(
+        "Client shouldn't have sent metadata twice."
+      )
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      try self.invalidState(
+        "Client can't have sent metadata if closed."
+      )
+    }
+  }
+
+  private mutating func serverReceive(bytes: ByteBuffer, endStream: Bool) throws {
+    switch self.state {
+    case .clientIdleServerIdle:
+      try self.invalidState(
+        "Can't have received a message if client is idle."
+      )
+    case .clientOpenServerIdle(var state):
+      // Deframer must be present on the server side, as we know the decompression
+      // algorithm from the moment the client opens.
+      try state.deframer!.process(buffer: bytes) { deframedMessage in
+        state.inboundMessageBuffer.append(deframedMessage)
+      }
+
+      if endStream {
+        self.state = .clientClosedServerIdle(.init(previousState: state))
+      } else {
+        self.state = .clientOpenServerIdle(state)
+      }
+    case .clientOpenServerOpen(var state):
+      try state.deframer.process(buffer: bytes) { deframedMessage in
+        state.inboundMessageBuffer.append(deframedMessage)
+      }
+
+      if endStream {
+        self.state = .clientClosedServerOpen(.init(previousState: state))
+      } else {
+        self.state = .clientOpenServerOpen(state)
+      }
+    case .clientOpenServerClosed(let state):
+      // Client is not done sending request, but server has already closed.
+      // Ignore the rest of the request: do nothing, unless endStream is set,
+      // in which case close the client.
+      if endStream {
+        self.state = .clientClosedServerClosed(.init(previousState: state))
+      }
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      try self.invalidState(
+        "Client can't send a message if closed."
+      )
+    }
+  }
+
+  private mutating func serverNextOutboundMessage() throws -> OnNextOutboundMessage {
+    switch self.state {
+    case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+      try self.invalidState("Server is not open yet.")
+    case .clientOpenServerOpen(var state):
+      let response = try state.framer.next(compressor: state.compressor)
+      self.state = .clientOpenServerOpen(state)
+      return response.map { .sendMessage($0) } ?? .awaitMoreMessages
+    case .clientClosedServerOpen(var state):
+      let response = try state.framer.next(compressor: state.compressor)
+      self.state = .clientClosedServerOpen(state)
+      return response.map { .sendMessage($0) } ?? .awaitMoreMessages
+    case .clientOpenServerClosed(var state):
+      let response = try state.framer.next(compressor: state.compressor)
+      if let response {
+        self.state = .clientOpenServerClosed(state)
+        return .sendMessage(response)
+      } else {
+        self.state = .clientOpenServerClosed(state)
+        return .noMoreMessages
+      }
```

Comment:
nit: reduce duplication

```suggestion
      let response = try state.framer.next(compressor: state.compressor)
      self.state = .clientOpenServerClosed(state)
      if let response {
        return .sendMessage(response)
      } else {
        return .noMoreMessages
      }
```